// inter/AdminControl.js
var async = require('async');

var config = require('config');
var ec = config.ec;
var prop = config.prop;

var cons = require('cons');
var keyFromType = cons.keyFromType;
var customerType = cons.customerType;

var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var log = esut.log;
var dateUtil = esut.dateUtil;

var dc = require('db').dc;

var Control = function () {
    var self = this;
    var cmdArray = [
        {code: 'AD01', fromType: keyFromType.CACHE, des: "新增游戏"},
        {code: 'AD02', fromType: keyFromType.CACHE, des: "更新游戏状态"},
        {code: 'AD03', fromType: keyFromType.CACHE, des: "新增游戏奖级"},
        {code: 'AD04', fromType: keyFromType.CACHE, des: "更新用户信息"},
        {code: 'AD05', fromType: keyFromType.CACHE, des: "新增商品信息"},
        {code: 'AD06', fromType: keyFromType.CACHE, des: "新增任务信息"}  //added by Xiao Xu
    ];
    self.cmd = {};
    for(var i = 0; i < cmdArray.length; i++)
    {
        var set = cmdArray[i];
        set.role = customerType.ADMIN.id;   //限制只有管理员可以访问
        self.cmd[set.code] = set;
    }
};

Control.prototype.checkAD01 = function (headNode, bodyNode, cb) {
    cb(null);
};

Control.prototype.checkAD02 = function (headNode, bodyNode, cb) {
    cb(null);
};

Control.prototype.checkAD03 = function (headNode, bodyNode, cb) {
    cb(null);
};

Control.prototype.checkAD04 = function (headNode, bodyNode, cb) {
    cb(null);
};

Control.prototype.checkAD05 = function (headNode, bodyNode, cb) {
    cb(null);
};

Control.prototype.checkAD06 = function (headNode, bodyNode, cb){
   cb(null);
};

/**
 * 新增游戏
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD01 = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    var book = {
        code:bodyNode.code,
        name:bodyNode.name,
        create_time:dateUtil.getCurTime()
    };
    async.waterfall([
        function(cb){   //保存到数据库
            var table = dc.pg.get("game");
            table.save(book, function(err, data){
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 更新游戏状态
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD02 = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    var cond = {
        id:bodyNode.id
    };
    var status = bodyNode.status;
    if(status > 0)
    {
        status = cons.gameStatus.OPEN.id;
    }
    else
    {
        status = cons.gameStatus.STOP.id;
    }
    var doc = {
        $set:{
            status:status
        }
    };
    async.waterfall([
        function(cb){   //保存到数据库
            var table = dc.pg.get("game");
            table.update(cond, doc, function(err, data){
                cb(err);
            })
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增游戏奖级
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD03 = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    var gl = {
        game_code:bodyNode.game_code,
        code:bodyNode.code,
        name:bodyNode.name,
        hit_rule:bodyNode.hit_rule,
        hit_amount:bodyNode.hit_amount,
        create_time:dateUtil.getCurTime()
    };
    async.waterfall([
        function(cb){   //保存到数据库
            var table = dc.pg.get("game_level");
            table.save(gl, function(err, data){
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 更新用户信息
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD04 = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    var cond = {id:bodyNode.customer_id};
    var docs = {
        $set: {
            status:bodyNode.status,
            password:bodyNode.password
        }
    }
    async.waterfall([
        function(cb){   //保存到数据库
            var table = dc.pg.get("customer");
            table.update(cond, docs, function(err, data){
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增商品信息
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD05 = function (headNode, bodyNode, cb) {
    var self = this;
    var backBodyNode = {};
    var goods = {
        name:bodyNode.name,
        num:bodyNode.num,
        amount:bodyNode.amount,
        remark:bodyNode.remark
    };
    async.waterfall([
        function(cb){   //保存到数据库
            var table = dc.pg.get("goods");
            table.save(goods, function(err, data){
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增任务信息
 * @param headNode
 * @param bodyNode
 * @param cb
 */
Control.prototype.handleAD06 = function (headNode, bodyNode, cb) {
    var self =  this;
    var backBodyNode = {};
    var task = {
        name:bodyNode.name,
        reward_exp:bodyNode.reward_exp,
        reward_gold:bodyNode.reward_gold,
        status:1
    };
    async.waterfall([
        function(cb){  // insert into the database
            var table = dc.pg.get("task");
            table.save(task, function(err, data){
                cb(err);
            });
        }
    ], function(err, result){
        cb(err, backBodyNode);
    });
};

module.exports = new Control();
