// admin/BookPageControl.js
var async = require('async');

var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;
var dateUtil = esut.dateUtil;

var service = require('service');
var loginService = service.loginService;

var config = require('config');
var ec = config.ec;

var dc = require('db').dc;

var cons = require('cons');
var customerType = cons.customerType;

var PageControl = function(){};

/**
 * 二层结构
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    if(self[cmd[2]])
    {
        self[cmd[2]](headNode, bodyNode, cb);
    }
    else
    {
        cb(ec[100]);
    }
};

PageControl.prototype.index = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 书籍类别管理
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.type = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增书籍类别
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.addType = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 书籍类别详情
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.detailType = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.parentId = bodyNode.parentId;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录
        {
            var table = dc.pg.get("book_type");
            var cond = {id:bodyNode.bookTypeId};
            table.find(cond, {}, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst[0];
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 查询书籍类别
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.listType = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;

    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("book_type");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(backBodyNode.cond, {}, options, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(backBodyNode.cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

var PageControl = new PageControl();
module.exports = PageControl;
