// admin/GamePageControl.js
var async = require('async');

var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;
var dateUtil = esut.dateUtil;

var service = require('service');
var loginService = service.loginService;

var config = require('config');
var ec = config.ec;

var dc = require('db').dc;

var cons = require('cons');
var customerType = cons.customerType;

var PageControl = function(){};

/**
 * 二层结构
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    if(self[cmd[2]])
    {
        self[cmd[2]](headNode, bodyNode, cb);
    }
    else
    {
        cb(ec[100]);
    }
};

/**
 * 游戏管理首页
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.index = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 查询所有
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.listGame = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;

    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("game");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(backBodyNode.cond, {}, options, function(err, data){
                if(data)
                {
                    for(var key in data.rst)
                    {
                        var set = data.rst[key];
                        set.status = cons.gameStatus.id[set.status];
                    }
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(backBodyNode.cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增游戏
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.addGame = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};


/**
 * 游戏奖级管理首页
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.manLevel = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增游戏等级
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.addGameLevel = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)
        {
            var table = dc.pg.get("game");
            table.find({}, {}, function(err, data){
                if(data)
                {
                    backBodyNode.game = data.rst;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 查询所有奖级
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.listGameLevel = function(headNode, bodyNode, cb)
{
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;

    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("game_level");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(backBodyNode.cond, {}, options, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(backBodyNode.cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

PageControl.prototype.listGoldPool = function(headNode, bodyNode, cb){
    var self = this;
    headNode.role = customerType.ADMIN.id;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;

    backBodyNode.sort = {id:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("goldPool");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(backBodyNode.cond, {}, options, function(err, data){
                if(data)
                {
                 /**   for(var key in data.rst)
                    {
                        var set = data.rst[key];
                        set.status = cons.gameStatus.id[set.status];
                    }*/
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(backBodyNode.cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

var PageControl = new PageControl();
module.exports = PageControl;
