// control/AdminPageControl.js
var config = require('config');
var ec = config.ec;

/**
 * admin只是一个中间层，因为管理后台页面有二层结构
 * @type {exports}
 */
var admin = require('admin');

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    if(self[cmd[1]])
    {
        self[cmd[1]](headNode, bodyNode, cb);
    }
    else
    {
        cb(ec[100]);
    }
};

/**
 * 用户管理
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.user = function(headNode, bodyNode, cb)
{
    admin.userCtl.handle(headNode, bodyNode, cb);
};

/**
 * 书籍管理
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.book = function(headNode, bodyNode, cb)
{
    admin.bookCtl.handle(headNode, bodyNode, cb);
};

/**
 * 后台管理，登录等公用的页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.sys = function(headNode, bodyNode, cb)
{
    admin.sysCtl.handle(headNode, bodyNode, cb);
};

/**
 * 游戏管理的页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.game = function(headNode, bodyNode, cb)
{
    admin.gameCtl.handle(headNode, bodyNode, cb);
};

/**
 * 投注管理的页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.ticket = function(headNode, bodyNode, cb)
{
    admin.ticketCtl.handle(headNode, bodyNode, cb);
};

/**
 * 帐户管理页面
 */
PageControl.prototype.account = function(headNode, bodyNode, cb)
{
    admin.accountCtl.handle(headNode, bodyNode, cb);
};

/**
 * 商品管理页面
 */
PageControl.prototype.goods = function(headNode, bodyNode, cb)
{
    admin.goodsCtl.handle(headNode, bodyNode, cb);
};

/**
 * 任务管理页面
 */
PageControl.prototype.task = function(headNode, bodyNode, cb)
{
    admin.taskCtl.handle(headNode, bodyNode, cb);
};

var PageControl = new PageControl();
module.exports = PageControl;
