// control/BookPageControl.js
var async = require('async');

var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;
var dateUtil = esut.dateUtil;

var service = require('service');
var loginService = service.loginService;

var config = require('config');
var ec = config.ec;


var dc = require('db').dc;

var cons = require('cons');
var fileType = cons.fileType;

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    if(self[cmd[1]])
    {
        self[cmd[1]](headNode, bodyNode, cb);
    }
    else
    {
        cb(ec[100]);
    }
};

PageControl.prototype.index = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    var userId = parseInt(headNode.userId);

    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("book");
    var cond = backBodyNode.cond;
    cond.customer_id = userId;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(cond, {}, options, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增书籍页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.add = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var table = dc.pg.get("book_type");
            table.find({}, {}, function(err, data){
                if(data)
                {
                    backBodyNode.bookType = data.rst;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 书籍详情页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.detail = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.parentId = bodyNode.parentId;
    backBodyNode.fileSite = cons.site.FILE.url;
    var userId = parseInt(headNode.userId);
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var table = dc.pg.get("book");
            var cond = {id:bodyNode.bookId, customer_id:userId};
            table.find(cond, {}, function(err, data){
                if(data && data.rst.length > 0)
                {
                    backBodyNode.book = data.rst[0];
                    cb(null);
                }
                else
                {
                    cb(ec[101]);
                }
            });
        },
        function(cb)    //获取类型信息
        {
            var table = dc.pg.get("book_type");
            var cond = {id:backBodyNode.book.book_type_id};
            table.find(cond, {}, function(err, data){
                if(data && data.rst.length > 0)
                {
                    backBodyNode.book.type = data.rst[0];
                    cb(null);
                }
                else
                {
                    cb(ec[101]);
                }
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 书籍章节列表
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.chapterList = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.bookId = bodyNode.bookId;
    var userId = parseInt(headNode.userId);

    backBodyNode.sort = {seq:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("chapter");
    var cond = backBodyNode.cond;
    cond.customer_id = userId;
    cond.book_id = bodyNode.bookId;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(cond, {}, options, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 书籍小节列表
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.blockList = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.chapter_id = bodyNode.chapter_id;
    backBodyNode.book_id = bodyNode.book_id;
    var userId = parseInt(headNode.userId);

    backBodyNode.sort = {seq:-1};
    backBodyNode.limit = 8;
    pageUtil.parse(bodyNode, backBodyNode);

    var table = dc.pg.get("chapter_block");
    var cond = backBodyNode.cond;
    cond.customer_id = userId;
    cond.chapter_id = bodyNode.chapter_id;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(cond, {}, options, function(err, data){
                if(data)
                {
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增章节
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.addChapter = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.book_id = bodyNode.book_id;
    var userId = parseInt(headNode.userId);

    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 新增文件块
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.addBlock = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.chapter_id = bodyNode.chapter_id;
    backBodyNode.book_id = bodyNode.book_id;
    var userId = parseInt(headNode.userId);

    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 小节编辑页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.detailBlock = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.block_id = bodyNode.block_id;
    var userId = parseInt(headNode.userId);

    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取用户的书本记录集
        {
            var table = dc.pg.get("chapter_block");
            table.find({id:backBodyNode.block_id}, {}, function(err, data){
                if(data && data.rst.length > 0)
                {
                    backBodyNode.block = data.rst[0];
                    cb(null);
                }
                else
                {
                    cb(ec[101]);
                }
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

var PageControl = new PageControl();
module.exports = PageControl;
