// control/PageControl.js
var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;

var testPageCtl = require("./TestPageControl.js");
var userPageCtrl = require("./UserPageControl.js");
var bookPageCtrl = require("./BookPageControl.js");
var adminPageCtrl = require("./AdminPageControl.js");
var sysPageCtrl = require('./SysPageControl.js');

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    console.log(bodyNode);
    var self = this;
    var cmd = headNode.cmd;
    if(self[cmd[0]])
    {
        self[cmd[0]](headNode, bodyNode, cb);
    }
    else
    {
        log.error("页面不存在!");
        log.error(cmd);
        cb("page not found!");
    }
};

PageControl.prototype.test = function(headNode, bodyNode, cb)
{
    testPageCtl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.user = function(headNode, bodyNode, cb)
{
    userPageCtrl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.book = function(headNode, bodyNode, cb)
{
    bookPageCtrl.handle(headNode, bodyNode, cb);
};

/**
 * 管理员页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.admin = function(headNode, bodyNode, cb)
{
    adminPageCtrl.handle(headNode, bodyNode, cb);
};

/**
 * 系统公用页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.sys = function(headNode, bodyNode, cb)
{
    sysPageCtrl.handle(headNode, bodyNode, cb);
};

var pageControl = new PageControl();
module.exports = pageControl;
