// control/TestPageControl.js
var async = require('async');

var esut = require('easy_util');
var digestUtil = esut.digestUtil;


var TestPageControl = function(){};

TestPageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[1]](headNode, bodyNode, cb);
};


TestPageControl.prototype.index = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.load = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.anim = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.define = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.tree = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.comp = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.empty = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};


TestPageControl.prototype.button = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.checkbox = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

TestPageControl.prototype.collapse = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

var testPageControl = new TestPageControl();
module.exports = testPageControl;
