// control/UserPageControl.js
var async = require('async');

var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;
var dateUtil = esut.dateUtil;

var service = require('service');
var loginService = service.loginService;

var config = require('config');
var ec = config.ec;

var dc = require('db').dc;

var cons = require('cons');
var fileType = cons.fileType;

var user = require('user');

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[1]](headNode, bodyNode, cb);
};

PageControl.prototype.register = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

PageControl.prototype.login = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

/**
 * 文件管理模块
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.file = function(headNode, bodyNode, cb)
{
    user.fileCtl.handle(headNode, bodyNode, cb);
};

/**
 * 票据管理模块
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.ticket = function(headNode, bodyNode, cb)
{
    user.ticketCtl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.index = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = digestUtil.createUUID();
    backBodyNode.parent = '';
    if(bodyNode.toUrl)  //要跳转到的目标页面
    {
        backBodyNode.toUrl = bodyNode.toUrl;
    }
    else
    {
        backBodyNode.toUrl = '';
    }
    backBodyNode.data = bodyNode.data;

    async.waterfall([
        function(cb){
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

PageControl.prototype.myInfo = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.fileSite = cons.site['FILE'].url;
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)
        {
            var table = dc.pg.get("customer");
            var cond = {
                id:parseInt(headNode.userId)
            }
            table.find(cond, {}, function(err, data){
                if(err)
                {
                    cb(ec[9999]);
                }
                else
                {
                    var customer = data.rst[0];
                    customer.password = customer.password.substr(0, 2) + '****';
                    backBodyNode.customer = customer;
                    cb(null);
                }
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 我的文件页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.myFile = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 10;
    pageUtil.parse(bodyNode, backBodyNode);
    console.log(backBodyNode);
    backBodyNode.fileType = fileType;

    var userId = parseInt(headNode.userId);
    var cond = backBodyNode.cond;
    cond.customer_id = userId;
    var table = dc.pg.get("customer_file");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(cond, {}, options, function(err, data){
                if(data)
                {
                    for(var key in data.rst)
                    {
                        var set = data.rst[key];
                        set.type = fileType.id[set.type];
                    }
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

/**
 * 文件管理页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.manFile = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)
        {
            cb(null);
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

PageControl.prototype.upload = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.parent = bodyNode.parent;
    cb(null, backBodyNode);
};

/**
 * 用户上传头像页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.uploadAvatar = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.loadIas = bodyNode.loadIas;
    cb(null, backBodyNode);
};

/**
 * 用户选择头像页面
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PageControl.prototype.selectImg = function(headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    backBodyNode.id = bodyNode.id;
    backBodyNode.sort = {create_time:-1};
    backBodyNode.limit = 4;
    pageUtil.parse(bodyNode, backBodyNode);
    backBodyNode.fileType = fileType;
    backBodyNode.fileSite = cons.site['FILE'].url;

    var userId = parseInt(headNode.userId);
    var cond = backBodyNode.cond;
    cond.customer_id = userId;
    cond.type = {$in:[fileType['JPEG'].id, fileType['PNG'].id]};
    var table = dc.pg.get("customer_file");
    async.waterfall([
        function(cb){   //校验登录
            loginService.pageCheck(headNode, function(err, data){
                cb(err);
            })
        },
        function(cb)    //获取记录集
        {
            var options = {sort:backBodyNode.sort, limit:backBodyNode.limit, offset:backBodyNode.skip};
            table.find(cond, {}, options, function(err, data){
                if(data)
                {
                    for(var key in data.rst)
                    {
                        var set = data.rst[key];
                        set.type = fileType.id[set.type];
                    }
                    backBodyNode.rst = data.rst;
                }
                cb(err);
            });
        },
        function(cb)    //获取总记录数目
        {
            table.count(cond, function(err, data){
                if(data)
                {
                    backBodyNode.count = data.rst[0].count;
                }
                cb(err);
            });
        }
    ], function (err, result) {
        cb(err, backBodyNode);
    });
};

PageControl.prototype.download = function(headNode, bodyNode, cb)
{
    var self = this;
    cb(null, {});
};

var PageControl = new PageControl();
module.exports = PageControl;
